### PR TITLE
test(sns): Add unit tests for automatic upgrades triggered when target version ahead of deployed version

### PR DESCRIPTION
**Pull request by anchpop**
This PR contributes unit tests of the whole automatic-upgrade process (as triggered by target_version being set). It adds a test for Governance and Root.

## Why add unit tests when we already have integration tests?

Certain scenarios are much easier to test via unit testing. These tests lay the groundwork for a test I want to write which will exercise edge cases w.r.t. archive canisters, which are too hard to exercise in an integration test

### CODE REVIEW

File: rs/sns/governance/src/governance/assorted_governance_tests.rs [hunk 1]
```diff
@@ -165,12 +165,12 @@ lazy_static! {
 
     pub(crate) static ref TEST_ROOT_CANISTER_ID: CanisterId = CanisterId::from(500);
     pub(crate) static ref TEST_GOVERNANCE_CANISTER_ID: CanisterId = CanisterId::from(501);
-    static ref TEST_LEDGER_CANISTER_ID: CanisterId = CanisterId::from(502);
-    static ref TEST_SWAP_CANISTER_ID: CanisterId = CanisterId::from(503);
-    static ref TEST_ARCHIVES_CANISTER_IDS: Vec<CanisterId> =
+    pub(crate) static ref TEST_LEDGER_CANISTER_ID: CanisterId = CanisterId::from(502);
```
Comment (max-dfinity) [inline rs/sns/governance/src/governance/assorted_governance_tests.rs:168]: I would move these so they're accessible in both, ideally, since it seems like we are finding a need for some common test code / fixtures.

File: rs/sns/governance/src/governance/advance_target_sns_version_tests.rs [hunk 2]
```diff
@@ -351,3 +635,177 @@ fn test_perform_advance_target_version() {
         assert_eq!(result, expected_result, "{}", label);
     }
 }
+
+fn add_calls_for_initiate_upgrade(
```
Comment (max-dfinity) [inline rs/sns/governance/src/governance/advance_target_sns_version_tests.rs:639]: set_environment_call_expectations_and_responses_for_initiate_upgrade

File: rs/sns/governance/src/governance/assorted_governance_tests.rs (hunk 1 above)
Comment (anchpop) [reply rs/sns/governance/src/governance/assorted_governance_tests.rs]: Done!

File: rs/sns/governance/src/governance/advance_target_sns_version_tests.rs [hunk 3]
```diff
@@ -351,3 +635,177 @@ fn test_perform_advance_target_version() {
...
+                    sender_canister_version: None,
+                })
+                .unwrap(),
+                Some(Ok(vec![])),
+            );
+        }
+    }
+}
+
+fn sns_canisters_summary(env: &mut NativeEnvironment, version: SnsVersion) {
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/governance/src/governance/advance_target_sns_version_tests.rs]: done!

File: rs/sns/governance/src/governance/advance_target_sns_version_tests.rs [hunk 4]
```diff
@@ -351,3 +635,177 @@ fn test_perform_advance_target_version() {
...
+            env.require_call_canister_invocation(
+                CanisterId::ic_00(),
+                "install_code",
+                Encode!(&ic_management_canister_types::InstallCodeArgs {
+                    mode: ic_management_canister_types::CanisterInstallMode::Upgrade,
+                    canister_id: canister_id.get(),
+                    wasm_module: vec![9, 8, 7, 6, 5, 4, 3, 2],
+                    arg: Encode!().unwrap(),
+                    compute_allocation: None,
+                    memory_allocation: None, // local const in install_code()
```
[the remark this answers is not in the document]
Comment (anchpop) [reply rs/sns/governance/src/governance/advance_target_sns_version_tests.rs]: done!

File: rs/sns/governance/src/governance/advance_target_sns_version_tests.rs (hunk 2 above)
Comment (anchpop) [reply rs/sns/governance/src/governance/advance_target_sns_version_tests.rs]: changed to somethign else!
